### PR TITLE
refactor(store): config-validator

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

Can override isAngularInTestMode

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/store/src/symbols.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ import { DispatchOutsideZoneNgxsExecutionStrategy } from './execution/dispatch-o
 export const ROOT_STATE_TOKEN = new InjectionToken<any>('ROOT_STATE_TOKEN');
 export const FEATURE_STATE_TOKEN = new InjectionToken<any>('FEATURE_STATE_TOKEN');
 export const NGXS_PLUGINS = new InjectionToken('NGXS_PLUGINS');
+export const NGXS_DEV_MODE = new InjectionToken('NGXS_DEV_MODE');
```
[the remark this answers is not in the document]
Comment (splincode) [reply packages/store/src/symbols.ts]: done

File: packages/store/src/module.ts [hunk 2]
```diff
@@ -60,6 +60,10 @@ export class NgxsModule {
         SelectFactory,
         PluginManager,
         ...states,
+        {
+          provide: NGXS_DEV_MODE,
+          useFactory: NgxsModule.isAngularInTestMode
```
[the remark this answers is not in the document]
Comment (splincode) [reply packages/store/src/module.ts]: done

File: packages/store/src/module.ts [hunk 3]
```diff
@@ -110,6 +92,45 @@ export class NgxsModule {
...
+    options: NgxsModuleOptions
+  ): Provider[] {
+    return [
+      {
+        provide: NGXS_TEST_MODE,
+        useFactory: NgxsModule.isAngularInTestMode
+      },
+      {
+        provide: NG_TEST_MODE,
+        useFactory: NgxsModule.isAngularDevMode
```
Comment (markwhitfeld) [inline packages/store/src/module.ts]: Surely this should be NG_DEV_MODE to match what it represents?

Comment (splincode) [reply packages/store/src/module.ts]: done

File: packages/store/src/internal/validation/config-validator.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import {
+  CONFIG_MESSAGES as MESSAGES,
+  VALIDATION_CODE as CODE
+} from '../validation/validation-messages.config';
+import { NG_DEV_MODE, NGXS_TEST_MODE, NgxsConfig } from '../../symbols';
+
+@Injectable()
+export class ConfigValidator {
+  constructor(
+    @Inject(NGXS_TEST_MODE) public isTestMode: Function,
```
Comment (markwhitfeld) [inline packages/store/src/internal/validation/config-validator.ts]: Sorry, I didn't notice this before. This is based on the mode of Angular, not NGXS, so it should rather be called `NG_TEST_MODE`.
So we would be injecting `NG_TEST_MODE` and `NG_DEV_MODE` tokens.... hmmm... they seem to go together...
Thinking about it now, how about we have an class called `HostEnvironment` and we have `isTestMode` and `isDevMode` methods on it?

File: packages/store/src/internal/validation/config-validator.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    @Inject(NG_DEV_MODE) public isNgDevMode: Function,
+    private _config: NgxsConfig
+  ) {}
+
+  private get isIncorrectProduction(): boolean {
+    return !this.isTestMode() && !this.isNgDevMode() && this._config.developmentMode;
```
Comment (markwhitfeld) [inline packages/store/src/internal/validation/config-validator.ts]: I didn't see this change before either.
Could you move the `!this.isTestMode()` back to being a guard clause as the first line in `verifyDevMode`?

File: packages/store/src/internal/validation/config-validator.ts (hunk 4 above)
Comment (splincode) [reply packages/store/src/internal/validation/config-validator.ts]: good idea

File: packages/store/src/internal/validation/config-validator.ts (hunk 5 above)
Comment (splincode) [reply packages/store/src/internal/validation/config-validator.ts]: ok

Comment (splincode) [reply packages/store/src/internal/validation/config-validator.ts]: done

File: packages/store/src/internal/validation/config-validator.ts (hunk 4 above)
Comment (splincode) [reply packages/store/src/internal/validation/config-validator.ts]: done

**Comment by markwhitfeld** [comment]
@splincode I see that the branch that this was committed to was called `bugfix/test-mode` but the PR is listed as a refactoring. Was there a bug fix in this work?
I'm happy for you to merge, but if there was a bug fix here then it should be described in the commit.

**Comment by markwhitfeld** [comment]
Sorry if I created confusion with my question.
Closed the other duplicate branch and merged this one because I would rather keep the conversation.

**Pull request merged**